HFR frontend rework

esainane opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Massive overhaul of the HFR frontend.

![hfr-ui-redux-8](https://user-images.githubusercontent.com/14355175/136348022-3dfe794c-0b17-4b95-834f-aa9333e1962b.png)

I'm still getting used to the tgui framework, so any guidance toward becoming mergeworthy would be appreciated.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

This makes the interface much more usable, and tries to help introduce the concepts involved in a very complicated system. The current interface is a bit clunky to use, and does not offer much guidance as to what actions are possible, or what the implications of taking them might be. You can't see the readouts at the same time you're tweaking the controls. Many failure states are not warned about, or even made clear when they're actively happening. To this end, this PR provides many very helpful qol changes, listed in the changelog below.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. -->

:cl:
code: LabeledList.Item now supports arbitrary nodes as labels. String labels still work as previously.
code: RoundGauge now supports an alertBefore property, for when warnings need to happen below a threshold. This can combine sensibly to form ranges with alertAfter.
refactor: The HFR interface has been heavily rewritten and restructured, fixing many edge cases where data was empty, unset, or took extreme values. Major parts now have their own file. The structure of gas lists is now consistent. Static data is used for more values. Much presentation layer logic has been moved to tgui instead of dm. Some processing that should never have happened, in tgui or anywhere else, is now gone forever.
qol: The HFR UI now displays the recipe specific gases that each fuel will create in a large table inside an expandable area. This can be collapsed when not needed.
qol: The HFR UI now uses a minimum mass for the scale of its gas lists. This should provide a visual cue for when a gas mix is running out, or needs more fuel to start.
qol: The HFR UI now graphs its various gas mixtures on a unified logarithmic scale, and tells you how fast the temperatures are changing in real units after all effects.
qol: The HFR UI now displays reactor status through gauges, with scales set to indicate typical ranges involved, and warnings to indicate when something is going wrong.
qol: The HFR UI now directly displays reaction instability, reaction activity, and the cell charge of the local APC. Reaction activity is the amount of heat output relative to what is permitted by the heat output limiter. This is effectively making explicit the result of a calculation you would otherwise need to do anyway, and makes it easier to notice when a reaction has stalled.
qol; The HFR UI will now helpfully inform you in discerning red text if you have no recipe set, or if any one of the gas mixtures is empty. This should be particularly helpful for noticing when you didn't add any coolant. Please involve coolant when operating a nuclear reactor.
qol: The HFR UI now always lists Plasma, BZ, and Proto-Nitrate in the moderator gas list, and provides tooltips explaining their usage and purpose.
qol: The HFR UI now has many other tooltips, from explaining the function and purpose of the various reactor controls, to the fringe effects of uncommon moderator gases.
code: heat_output_bool is gone forever yes this deserves its own changelog entry
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

esainane committed:
HFR UI Rework: Early experiments: Scratch ideas

esainane committed:
HFR UI Rework: Early experiments: Just use table

esainane committed:
HFR UI Rework 3: Monolith edition

No more tabs

esainane committed:
LabeledList.Item: Handle InfernoNode labels

This allows the optional rendering of non-string items as the label in a
LabeledList.Item.

esainane committed:
RoundGauge: Add alertBefore prop

This is a property similar to alertAfter, but displays an alert below
a threshold. If both thresholds are set, and define a range between
them, the alert will only flash when the needle is between both
values. Otherwise, the alert will flash when on the active side of
either threshold.

esainane committed:
HFR UI: Mass rewrite and restructure

The UI is now split into multiple files, at present still tightly coupled.
Where possible, data passing is made pure and explicit rather than passing
the full data object around.

The Controls file still references useBackend.

This depends on improvements made to tgui components previously.

esainane committed:
HFR UI: Energy: Prevent -Infinity

It still gets clamped to the range (10**12 considered infintesimal),
but this stops the label from being silly.

esainane committed:
HFR UI: Energy: Set minimum prefix to T

If you manage to drop to GJ on the HFR, it's effectively zero anyway.

esainane committed:
HFR UI: Energy: Less magic numbers

Also make the black band explicit.

esainane committed:
HFR UI: Nicer boundaries for Fusion Level Gauge

This has the needle point at the middle of a band for eg Fusion Level 4,
which makes it much clearer where which state of danger begins and ends.

esainane committed:
HFR UI: Gases: Higher minimum scale

This makes it easier to convey required gas amounts and to notice
low gas conditions.

esainane committed:
HFR UI: Temperatures: key-ify

esainane committed:
HFR UI: Temperatures: Exact value tooltip

esainane committed:
HFR UI: Adjust automatic spacing

Some of these properties should be passed on from the base layout.

esainane committed:
HFR UI: Recipes: Fix pure check

This drastically reduces performance load when visible.

esainane committed:
HFR UI: Alert at power level 5 too

esainane committed:
HFR UI: Avoid graph render when idle

esainane committed:
HFR UI: Bars and controls

- Kill the horror that was heat_output_bool
- Expose instability to the UI
- Expose heat output minimums and maximums to the UI
- Expose preview heat values to the UI
- Kill the fusion temperature delta bar
- Bars now display current temperature and delta per sec on UI
- If an air is empty, it is displayed as "Empty" instead, as red warning
- Heat Change/Cap gauge is gone, too densely packed and hard to read.
- Reaction activity is now displayed directly
- Instability is now displayed directly
- Gauges in states blatantly not suitable to be left unattended now
  color their arc in yellow (non-minimized heating conductor when
  exothermic, active current damper)
- Fusion Level gauge is now even larger

esainane committed:
HFR UI: Fix negative notation

esainane committed:
HFR UI: Display APC energy

esainane committed:
HFR UI: Mass delint

esainane committed:
HFR UI: Rearrange interface

Move Act* into helpers

Factor out Fusion Gas/Moderator Gas commonalities to GasList and add
helptext

esainane committed:
HFR UI: More help text

esainane committed:
HFR UI: Zero out activity when off

esainane committed:
HFR UI: Logarithmic temperature scale

esainane committed:
HFR UI: Better help text

esainane committed:
HFR UI: Remove ActNone

esainane committed:
HFR UI: Unify Hypertorus subdir

esainane committed:
HFR UI: Local data extraction

Seeing the flow of data was useful during development, but unnecessary
and only adding clutter going forward.

esainane committed:
HFR UI: Cleanup, documentation

esainane committed:
HFR UI: CSSify, cleanup

esainane committed:
HFR UI: Mass delint 2

esainane committed:
HFR UI: Remove MaybeTooltip

Tooltip performance is fixed for real now, probably

Ghilker commented:
Great work as always

esainane commented on code/modules/atmospherics/machinery/components/fusion/hfr_core.dm:
```diff
@@ -144,10 +148,16 @@
+	var/output_temperature_archived = 0
 	var/output_temperature = 0
+	///Time between current and _archived temperatures
+	var/temperature_period = 1
```
It's constant here, but this will store the `delta_time` passed in to `process` after the backend rework.

tralezab commented:
![image](https://user-images.githubusercontent.com/40974010/134992785-c187a09f-460e-4371-b4b4-a5ceb7e97f3f.png)
i don't like this really, and the dead space at the bottom. otherwise looks good

esainane committed:
HFR: Use area APC cache

Ghilker commented on tgui/packages/tgui/interfaces/Hypertorus/Recipes.js:
```diff
@@ -0,0 +1,220 @@
+    icon: "thermometer-full",
+    override_base: 0.85,
+    scale: 1.15,
+    tooltip: (v, d) => "Maximum: " + (d.baseMaximumTemperature * v).toExponential() + "°K",
```
```suggestion
    tooltip: (v, d) => "Maximum: " + (d.baseMaximumTemperature * v).toExponential() + "K",
```
AHHHH

Ghilker commented on tgui/packages/tgui/interfaces/Hypertorus/Temperatures.js:
```diff
@@ -0,0 +1,226 @@
+      {value > 0
+        ? (
+          <>
+            <Box align="center">{to_exponential_if_big(value) + "°K"}</Box>
```
```suggestion
            <Box align="center">{to_exponential_if_big(value) + "K"}</Box>
```
More

Ghilker commented on tgui/packages/tgui/interfaces/Hypertorus/Temperatures.js:
```diff
@@ -0,0 +1,226 @@
+            <Box align="center">{to_exponential_if_big(value) + "°K"}</Box>
+            <Box align="center">{(delta === 0
+              ? '-'
+              : `${delta < 0 ? '' : '+'}${to_exponential_if_big(delta)}°K/s`
```
```suggestion
              : `${delta < 0 ? '' : '+'}${to_exponential_if_big(delta)}K/s`
```
Another one

Ghilker commented on tgui/packages/tgui/interfaces/Hypertorus/Temperatures.js:
```diff
@@ -0,0 +1,226 @@
+            className="hypertorus-temperatures__y-axis-label-icon"
+            name={icon}
+          />)}
+        {to_exponential_if_big(value) + "°K"}
```
```suggestion
        {to_exponential_if_big(value) + "K"}
```
More

stylemistake commented:
Overall good, code formatting could be better.

esainane committed:
HFR UI: Sort imports, remove blank lines

esainane committed:
HFR UI: Remove Act*

ActParam is useless since the introduction of ComputedPropertyName, and
the helpers that wrapped it no longer do enough to justify their existence.

LemonInTheDark commented:
@stylemistake are you happy with this? or is it just the volume of code involved

esainane committed:
HFR UI: Temperature graph readability

This was a pretty big mess. Split things up to use meaningfully named
constants and functions.

esainane committed:
HFR UI: span -> div

Deactivate 1990s DHTML era neuron

esainane committed:
HFR UI: Don't pass around act for ComboKnob

esainane committed:
HFR UI: Smarter window layout/flow

Each section now has an ideal width, and will try to fill it where
possible. Sections which can stack well with one another are nested
in further Flex layouts to save space in extreme scenarios, such as
if you really want to make the window extremely wide or extremely thin.

Sections that want to be wide, but have content that can collapse if need
be, also now have their content nested in a Flex layout.

Sections which really can't be reflowed easily - the graph and the recipes
table - are given a horizontal scroll bar to make the best out of the
situation.

The interface is still quite usable at 420px wide. We stay technically
functional at 300px wide, with annoying horizontal scroll bars,
but anything narrower will start to crop the gas quantity lists.

For comparison, R&D Console starts with a width of 640px, and starts
overlapping text around 470px.

esainane committed:
HFR UI: Relint
